### DISCUSSION TITLE
>>> Runtime error calling function: load_lib <<<

### DISCUSSION THREAD

**Discussion by ghost** [objeck/objeck-lang#226] [General]
This is the first time I try to write Objeck's JNI.

foo.h

```
#ifndef FOO_H
#define FOO_H

#include "../../vm/lib_api.h"
#include "../../shared/sys.h"

#ifdef _WIN32
#define DLL_EXPORT __declspec(dllexport) 
#endif

int foo(int a, int b);

#endif

```

foo.cpp

```
#include "foo.h"

int foo(int a, int b)
{
	return a + b;
}

extern "C" {

DLL_EXPORT void Foo(VMContext& context)
{
	const auto a = APITools_GetIntValue(context, 1);
	const auto b = APITools_GetIntValue(context, 2);
	APITools_SetIntValue(context, 0, foo(a, b));
}

}

```

foo.obs

```
#~
Library foo (-lib foo)
~#

use System.API;

bundle Foo {
	class Proxy {
		@lib_proxy : static : DllProxy;
		
		function : GetDllProxy() ~ DllProxy {
			if(@lib_proxy = Nil) {
				@lib_proxy := DllProxy->New("libobjk_foo");
			};

			return @lib_proxy;
		}
	}

	class TestFoo {
		@x : Int;
		@y : Int;
	
		New(x : Int, y : Int) {
			@x := x;
			@y := y;
		}
		
		method : public : TestFoo() ~ Int {
			array_args := Base->New[3];
			array_args[1] := IntRef->New(@x);
			array_args[2] := IntRef->New(@y);
			@lib_proxy := Proxy->GetDllProxy();
			@lib_proxy->CallFunction("Foo", array_args);
			value := array_args[0]->As(IntRef)
			return value->Get();
		}
	}
}

```

test.obs

```
use Foo;

class Test {
   function : Main(args : String[]) ~ Nil {
      hiya := TestFoo->New(1,2);
      hiya->TestFoo()->PrintLine();
   }
}

```

p/s: I have copied foo.obl and libobjk_foo.dll to their proper locations.

**Comment by objeck** [comment]
Are you not setting the return value? Try ```array_args[0] := IntRef->New();```

**Reply by ghost** [reply]
Did as you said. That doesn't make any differences.

**Reply by ghost** [reply]
@objeck Please test if it works for you and tell me if I did anything wrong. Thank you.

**Reply by objeck** [reply]
When I have a minute, I will create an example. What environment are you running under?

**Reply by ghost** [reply]
MinGW. Standalone MinGW, not msys2: https://github.com/mstorsjo/llvm-mingw

**Comment by objeck** [comment]
Please refer to [this](https://github.com/objeck/objeck-lang/tree/master/programs/frameworks/native) code example.

**Comment by ghost** [comment]
Why I'm wasting time with this language?

This is the command I used to compile the library:

`obc -src foo.obs -tar lib -dest foo.obl`

It always tells me `foo.obs:(2,2): Libraries and web applications may not define a 'Main(args)' function` and failed to generate any .obl file.

I tried to compile it as normal program with `obc -src foo.obs` and it worked and gave me a .obe file. But I want a .obl file, not .obe!

The command I used to compile a library is exactly like what the document tells me!

https://objeck.org/getting_started.html#create_libs

p/s: Thank you for your help but I think this language is not for me!

**Comment by objeck** [comment]
In the example above, you are building a library that should not have a ``Main`` function. Once a library is created and copied to the library directory, it can be included in a program using ``-lib xyz.obl``.

**Reply by ghost** [reply]
Please look at the command he used, I think he already used the correct command to build a library.

**Reply by ghost** [reply]
On https://objeck.org/getting_started.html#create_libs did you find something abnormal?

`obc -src pair.obs –tar lib -dest pair.obl`

`obc -src points.obs –lib pair`

Shouldn't it be `obc -src pair.obs -tar lib -dest pair.obl` and `obc -src points.obs -lib pair`? Did you notice the differences?

If someone copy/paste the commands from the document, I think it's very likely it will not work.

**Reply by ghost** [reply]
For unknown reasons, `obc` will not complain about the invalid switches (`–tar` and `–lib`). `–` on UTF-8 is `E2 80 93` when `-` is `2D`.

**Reply by ghost** [reply]
Nope, I think he used the correct command @objeck. I tried with `obc -src foo.obs -tar lib -dest foo.obl` and `obc` tell me the same error:

`foo.obs:(2,2): Libraries and web applications may not define a 'Main(args)' function`

This is definitely a bug.

**Reply by ghost** [reply]
Note: I use the release provided by you, not my own MSYS2 builds.

**Reply by objeck** [reply]
What code are you trying to compile into a library? Libraries cannot contain a Main function.
